### PR TITLE
feat: VSCode/(neo)vim Autocomplete

### PR DESCRIPTION
**Pull request by knownasilya**
Screen Shot 2019-12-02 at 2 51 59 PM

Adds autocomplete when used with unstable ember language server.
Should also work with Vim and Neovim

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,12 +69,20 @@
...
   "ember-addon": {
     "configPath": "tests/dummy/config",
     "before": [
       "ember-cli-htmlbars"
     ]
+  },
+  "ember-language-server": {
+    "entry": "./lib/langserver.js",
+    "version": 1,
+    "debug" : true,
```
Comment (lifeart) [inline package.json]: should be false for prod mode

File: lib/langserver.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+const { hasOctaneImports, transformOctaneImports } = require('./octane-utils');
+const { transformImports } = require('./utils');
+
+async function onComplete(projectRoot, opts) {
+  function log() {
+    return opts.server.connection.console.log(...arguments);
+  }
+
+  if (opts.focusPath.node.type === 'ElementNode') {
+    let filePath = uriToPath(opts.textDocument.uri);
```
Comment (lifeart) [inline lib/langserver.js]: const documentContent = opts.textDocument.getText();

File: lib/langserver.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      let content = await fs.readFile(filePath, 'utf-8');
+      let mightHaveImports = content.includes('import');
+      if (!mightHaveImports) {
+        return [];
+      }
+      let isOctane = hasOctaneImports(content);
+      let componentImports;
+
+      if (isOctane) {
+        let legacyContent = transformOctaneImports(content, filePath);
```
Comment (knownasilya) [inline lib/langserver.js]: Any way to get the file path without having to keep `uriToPath`? Looked at the source but didn't see anything like that

Comment (knownasilya) [reply lib/langserver.js]: @lifeart

Comment (lifeart) [reply lib/langserver.js]: yeah, you have to use it, I'm not confident about including it to public API, but it discussable

File: lib/langserver.js (hunk 2 above)
Comment (knownasilya) [reply lib/langserver.js]: hum, `textDocument` only has `uri` on it.

**Pull request merged**